### PR TITLE
[FSDP2] precompute scale after optimizer.step for dynamic scaling

### PR DESCRIPTION
**Pull request by weifengpy**
Goal: improve float8 all-gather perf in FSDP2 by precomputing scales for all float8 params with a single all-reduce

updated README for API usage: call `precompute_float8_scale_for_fsdp` inside the training loop after optimizer step

```
from float8_experimental.fsdp_utils import precompute_float8_amax_for_fsdp
# inside the training loop
model(input).sum().backward()
optim.step()
precompute_float8_dynamic_scale_for_fsdp(model)
```

unit test `pytest -s test/test_fsdp2/test_fsdp2_eager.py -k test_transformer_parity_dynamic`

**FSDP pre-forward**: shortend from 3ms to 1.8ms because of doing 1 all-reduce instead N small all-reduces
Screenshot 2024-05-30 at 12 38 24 AM

Screenshot 2024-05-30 at 12 48 14 AM

**Pre-computing amax**: shortened from 5ms to 1.7ms, by switching from `torch._foreach_abs` + `torch.max(a)` to `torch._foreach_norm(weights, ord=math.inf)`

Screenshot 2024-05-30 at 12 50 17 AM

Screenshot 2024-05-30 at 12 49 54 AM

### CODE REVIEW

File: float8_experimental/float8_dynamic_linear.py
Comment (bdhirsh) [inline float8_experimental/float8_dynamic_linear.py:157]: One major requirement for tensor subclasses that I don't think is respected here: `__tensor_flatten__` and `__tensor_unflatten__` must properly convey every inner tensor on the subclass.

So when we call `__tensor_flatten__` on this subclass, if either of `_fp8_data/scale/amax` are set to valid tensors, they need to be returned there (and similarly `__tensor_unflatten__` needs to handle them as extra args)

Comment (weifengpy) [reply float8_experimental/float8_dynamic_linear.py:157]: thanks for pointing this out! This saves me a lot of debugging time. I can give it a try by including `_fp8_data/scale/amax` in `__tensor_flatten__` and `__tensor_unflatten__`

File: float8_experimental/float8_linear_utils.py
Comment (weifengpy) [inline float8_experimental/float8_linear_utils.py:350]: `torch.clamp` calls `all_reduce`. since we performed `torch.clamp` here, I avoided calling it again in `amax_to_scale(clamp_amax=False)`

File: float8_experimental/float8_dynamic_linear.py
Comment (weifengpy) [inline float8_experimental/float8_dynamic_linear.py:193]: if `_pre_computed_amax`, we skip `tensor_to_amax` and directly do `amax_to_scale`

File: float8_experimental/float8_dynamic_linear.py
Comment (weifengpy) [reply float8_experimental/float8_dynamic_linear.py]: `torch.compile` works after patching https://github.com/pytorch/pytorch/pull/127431 
will compare traces in 2nd PR

File: float8_experimental/float8_linear_utils.py
Comment (awgu) [reply float8_experimental/float8_linear_utils.py:350]: So you are relying on `torch.clamp` to run the all-reduce implicitly from changing sharding from partial to replicate?

If this fragments the code, could we just all-reduce the amax tensor and then leave the clamp to `amax_to_scale`? I agree the current way is faster since we are doing one clamp for all amaxes, but in case float8 folks are not happy with this fragmentation, this seems like another way.

**Review by awgu — COMMENTED** [review]
This seems reasonable to me! I want to check with float8 folks on the `amax_to_scale` change.

File: float8_experimental/float8_linear_utils.py
Comment (weifengpy) [reply float8_experimental/float8_linear_utils.py:350]: thanks for the suggestions. I can collect feedback from float8 folks if they have a preference

**Comment by vkuzo** [comment]
nice! Can we include the intended user API in the PR summary?

File: float8_experimental/float8_dynamic_linear.py
Comment (drisspg) [reply float8_experimental/float8_dynamic_linear.py:154]: +1 on adding to flatten/unflatten

File: float8_experimental/float8_linear_utils.py
Comment (drisspg) [reply float8_experimental/float8_linear_utils.py:350]: agreed

File: float8_experimental/float8_utils.py
Comment (drisspg) [inline float8_experimental/float8_utils.py:44]: nit: I think if you have this on a seperate line
`amax = clamp(amax, eps) if clamp_amax else amax`

makes the logic a lil easier to follow

File: float8_experimental/float8_linear_utils.py
Comment (bdhirsh) [inline float8_experimental/float8_linear_utils.py:333]: @weifengpy do you plan / want to use compile on this, and are there any gaps around here that you think would be good to prioritize on the compile side?

This is mostly just me remembering @awgu mention a while ago that he thought compile added noticeable runtime overhead, and I can't remember if it was for this specific case. If it is, and we think compiling this code would be useful, I can prioritize looking into the runtime overhead.

Comment (weifengpy) [reply float8_experimental/float8_linear_utils.py:333]: Hi @bdhirsh, I plan to polish and land this PR without compile next week to conclude H1. most importantly add `_pre_computed_amax` to flatten/unflatten

Reducing runtime overhead from `torch.compile` is still meaningful since we want torch.compile(fp8 casting) in FSDP2 pre-forward hooks. would it be helpful if I work on a mini repro with profiler traces? Want to unblock you in the short-term

Comment (bdhirsh) [reply float8_experimental/float8_linear_utils.py:333]: If you have a mini repro showing bad runtime overheads with compile, that would be great!

Comment (weifengpy) [reply float8_experimental/float8_linear_utils.py:333]: Hi @bdhirsh , I have created a repro https://github.com/pytorch/pytorch/issues/129457 . I highlighted extra cpu overhead and gpu time for torch.compile(mode="reduce-overhead")

File: float8_experimental/float8_dynamic_utils.py
Comment (weifengpy) [inline float8_experimental/float8_dynamic_utils.py:146]: use if-else to handle optional tensor `_pre_computed_amax` otherwise dynamo will complain about `None`

File: float8_experimental/float8_dynamic_utils.py
Comment (weifengpy) [inline float8_experimental/float8_dynamic_utils.py:15]: pre-commit hook trigger linter, and cleaned unused import

File: float8_experimental/float8_linear_utils.py
Comment (weifengpy) [reply float8_experimental/float8_linear_utils.py:357]: moving to `fsdp_utils.py` according to PR https://github.com/pytorch-labs/float8_experimental/pull/310

Comment (weifengpy) [reply float8_experimental/float8_linear_utils.py]: indicating fsdp by renaming to `precompute_float8_amax_for_fsdp`

**Comment by weifengpy** [comment]
> nice! Can we include the intended user API in the PR summary?

added example API usage in PR summary

File: float8_experimental/float8_linear_utils.py
Comment (weifengpy) [reply float8_experimental/float8_linear_utils.py]: done

File: float8_experimental/float8_dynamic_linear.py
Comment (weifengpy) [reply float8_experimental/float8_dynamic_linear.py]: done

File: test/test_fsdp2/test_fsdp2_common.py
Comment (awgu) [inline test/test_fsdp2/test_fsdp2_common.py:9]: I might be very confused, but where is this `fsdp_utils` file?
Screenshot 2024-07-09 at 7 20 22 PM

and a nit: maybe should be consistent in whether `pre_compute` vs. `precompute` across the PR
(either way is fine to me, `precompute` maybe since it is more concise)

Comment (weifengpy) [reply test/test_fsdp2/test_fsdp2_common.py:9]: > I might be very confused, but where is this fsdp_utils file?

just added `float8_experimental/fsdp_utils.py`

> aybe should be consistent in whether pre_compute vs. precompute across the PR

good suggestion. I will use `precompute` consistently

File: float8_experimental/fsdp_utils.py
Comment (weifengpy) [inline float8_experimental/fsdp_utils.py:14]: improve docstring with example API usage

**Comment by weifengpy** [comment]
the linter error comes from trunk. I opened another PR to fix them: https://github.com/pytorch-labs/float8_experimental/pull/313

File: float8_experimental/float8_dynamic_utils.py
Comment (vkuzo) [inline float8_experimental/float8_dynamic_utils.py:150]: does having `Optional[torch.Tensor]` as a subclass field work with torch.compile? Or do we not care about torch.compile in this code path?

File: float8_experimental/fsdp_utils.py
Comment (vkuzo) [reply float8_experimental/fsdp_utils.py:14]: nice! can we add this to the README?

File: float8_experimental/float8_dynamic_utils.py
Comment (weifengpy) [reply float8_experimental/float8_dynamic_utils.py:150]: torch.compile assumes every tensor from `__tensor_flatten__` is not None. That's why I added `if-else` so when 
`_precomputed_amax is None`, it won't be included in `__tensor_flatten__`

with the `if-else`, torch.compile works. I verified it here and reported perf gaps with eager https://github.com/pytorch/pytorch/issues/129457

File: float8_experimental/fsdp_utils.py
Comment (weifengpy) [reply float8_experimental/fsdp_utils.py:14]: just added API usage to README

File: float8_experimental/float8_utils.py
Comment (weifengpy) [reply float8_experimental/float8_utils.py]: good suggestion! I changed the API to precompute scale and it shows another 9% speed up *in unit test* vs precomputing amax

`fsdp_pre_all_gather` is also greatly simplified because of using `self._precomputed_scale`

File: README.md
Comment (vkuzo) [inline README.md:61]: should we say that this is specific to FSDP2 with float8 all-gather turned on?  Also, maybe we can show how to turn that on, since I don't think it's documented in the README yet? Can be a follow-up PR.

File: float8_experimental/fsdp_utils.py
Comment (vkuzo) [reply float8_experimental/fsdp_utils.py:14]: nit: maybe we can make sure `dynamic` is in the name, since this is specific to dynamic scaling?

File: float8_experimental/fsdp_utils.py
Comment (vkuzo) [inline float8_experimental/fsdp_utils.py:24]: I think this can be relaxed to something like

```
if isinstance(m, Float8Linear) and m.scaling_type_w is TensorScalingType.DELAYED:
    raise NotImplementedError(...)
```

the scaling of activations and gradient's should not matter for this case

File: float8_experimental/fsdp_utils.py
Comment (vkuzo) [inline float8_experimental/fsdp_utils.py:32]: is this expensive for real models? if yes, maybe we can offer option to precompute this?

File: float8_experimental/fsdp_utils.py
Comment (vkuzo) [inline float8_experimental/fsdp_utils.py:41]: optional nit: maybe move outside to prevent nested functions?

**Review by vkuzo — APPROVED** [review]
looks great! had some final comments. thanks for doing this!

File: float8_experimental/fsdp_utils.py
Comment (awgu) [inline float8_experimental/fsdp_utils.py:13]: Should we add a `@torch.no_grad()` decorator on this?

File: float8_experimental/fsdp_utils.py
Comment (awgu) [inline float8_experimental/fsdp_utils.py:16]: suggestion:
```suggestion
    Calculate scale dynamically for all float8 parameters.
    This should be run after the optimizer step. It performs a single all-reduce to compute the
    amaxes for all float8 weights.
```

File: float8_experimental/fsdp_utils.py
Comment (awgu) [inline float8_experimental/fsdp_utils.py:17]: nit (typo):
```suggestion
    Example usage:
```

@vkuzo I assume that there are no docs builds for `float8_experimental`, so this example is for users who will read the code itself?

Otherwise, we might need to check the formatting -- I recall the format for examples being a bit different.

File: float8_experimental/fsdp_utils.py
Comment (awgu) [reply float8_experimental/fsdp_utils.py:32]: My intuition is that this should be pretty fast as the number of `nn.Module`s in the model is usually at most in the thousands and this is pure Python overhead. @weifengpy you can check the traces you have if you see any noticeable gaps from this.

File: float8_experimental/fsdp_utils.py
Comment (awgu) [reply float8_experimental/fsdp_utils.py:41]: curious what is the downside of nested functions

@weifengpy By the way, this was originally a nested function just so that we could try to `torch.compile` it effectively in the `scales = compute_scales(weights)` line. Does it still need to be a separate function for `torch.compile` reasons? If so, we should probably add a comment before the `def compute_scales` mentioning that it is separate for `torch.compile`; otherwise, we can consider inlining the function.

File: float8_experimental/fsdp_utils.py
Comment (awgu) [inline float8_experimental/fsdp_utils.py:60]: function name in the warning needs to be updated

I am okay with not including this warning by the way. This was also to help debugging to make sure we actually found `weights`.

File: float8_experimental/fsdp_utils.py
Comment (weifengpy) [reply float8_experimental/fsdp_utils.py:41]: I will remove nested functions to make the code easy to read. I profiled the unit test and `precompute_float8_scale_for_fsdp` takes 1.9ms. thta's a tiny portion of the overall training loop. No obvious reason to speed up with `torch.compile` yet. I can bring back the nested function in case we need `torch.compile` again.

Screenshot 2024-07-11 at 1 50 46 PM

File: README.md
Comment (weifengpy) [reply README.md:61]: > should we say that this is specific to FSDP2 with float8 all-gather turned on?

will change it in this PR

> maybe we can show how to turn that on, since I don't think it's documented in the README yet

good catch. will polish README again after landing changes in torchtitan to turn on/off fp8 all-gather

File: float8_experimental/fsdp_utils.py
Comment (weifengpy) [reply float8_experimental/fsdp_utils.py:13]: good catch. adding @torch.no_grad()

File: float8_experimental/fsdp_utils.py
Comment (weifengpy) [reply float8_experimental/fsdp_utils.py:14]: renaming to `precompute_float8_dynamic_scale_for_fsdp`

File: float8_experimental/fsdp_utils.py
Comment (weifengpy) [reply float8_experimental/fsdp_utils.py:60]: got you. I am removing the warnings for simplicity

File: float8_experimental/fsdp_utils.py
Comment (weifengpy) [reply float8_experimental/fsdp_utils.py:29]: just checked the profiler traces. it's roughly 0.15ms cpu overhead (5% of `precompute_float8_dynamic_scale_for_fsdp` and is tiny portion of 1 training loop). no cuda are launched

thus I am keeping it as is now for simplicity

**Pull request closed without merging**
